### PR TITLE
Improve nginx conf files

### LINKED ISSUE #9967: ♻️ Refactor nginx conf so we don't need to add an entry for every top-level path
## ♻️ Debt/Refactor

Currently our nginx conf files have a location block for every top-level path of the app. Any uri not covered by these paths gets routed to the the tc-report if possible, or a 404.

This seems backwards. There is a very limited number of tc-report paths, and we will never add more, whereas we will continue to add top-level paths to the main app. We should try to serve tc-report files first, fall back to the web app, and avoid as many location blocks as possible.

## 🕵️ Details

Inspired by discussion in the comments of #5186 

## 🙋‍♀️ Proposed Solution

Use [try_files](https://nginx.org/en/docs/http/ngx_http_core_module.html#try_files) to attempt to serve tc-report files first, then fall back to serving our webapp index file.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the debt was addressed and expected functionality has not been affected.

- [ ] Adding a new top-level path to the webapp does not require modifying the nginx conf
- [ ] Existing functionality in terms of routing and serving of files is preserved

### PR DESCRIPTION
**Pull request by tristan-orourke**
🤖 Resolves #9967

## 👋 Introduction

This simplifies the nginx conf without changing its functionality. 

## 🕵️ Details

Where it used to try to match known paths of the webapp, and then fall back to tc-report, it now tries to match tc-report files and asset files, and then falls back to the webapp index for all other urls. That means we can add new top-level paths to the web app without adding location blocks to the nginx conf! It also means a smaller, simpler file overall. 

I also moved things around a bit so that, for example, all permanent redirects are together. 

The one thing that is functionally different is a permanent redirect from `/privacy-notice` to `/privacy-policy`. This means that if you go to `talent-cloud` (a tc-report page) and click the "Privacy Policy" link in the footer, you will no longer end up on the old tc-report privacy page.

## 🧪 Testing

1. Restart your webserver to ensure nginx is using the new conf file. With a docker-compose setup, this can be done with `docker-compose restart webserver`
2. Open the browser inspect, go to network tab, and check "Disable cache"
3. Navigate through the site, either typing urls manually or refreshing in the browser, to avoid SPA navigation. Specific things to try:
    - urls work with or without a trailing slash
    - page urls (not including tc-report) work with or without  preceding language code
    - all asset files (images, js, css, webmanifests, etc) are loaded correctly
    - tc-report pages work (start by going to `/en/talent-cloud` and click around from there. Check with and without trailing slashes
    - `/config.js` returns a file
4. Now deploy this branch to the Dev vertical, to test the deploy conf file! Test the same things as above, and in addition, check things have the correct cache-control headers:
    - content-hashed files (eg most js files) should be cached for a year
    - non-content-hashed asset files (eg `/images/digital-talent/favicon.svg`) are cached for an hour
    - urls which look like an asset file but which don't exist (eg `/images/digital-talent/fake-favicon.svg`) return an actual 404 status code and redirect to display our 404 page
    - `/config.js` is cached for 1 minute
    - all tc-report content (html and asset files) are cached for a week
    - all other urls (real pages, and non-asset-like 404s) use `no-cache`

## 🧠 Other thoughts

- If we didn't mind caching all our asset files for the same amount of time, the location blocks could be simplified even more
- I'd love to add some prefix to all our asset file urls. I assume this could be done fairly easily with webpack? Then we could easily detect those files without using a regex, which would be slightly faster.
- Same goes for tc-report files, it would be nice to serve them all at urls prefixed with `/tc-report/`. We could do it if we add permenent redirects for the paths currently in the tc-report regex block.
- I considered moving all the permanent redirects to a separate file and using the `include` statement. This would keep the main file tidy, even if we want to add more and more permanent redirects (eg for tc-report). It could also be shared between the local and deploy configs, which are identical in that section. I skipped that for now to keep this PR simpler.

## 🚚 Deployment

Normal deployment scripts should be sufficient here.

### CODE REVIEW

**Review by petertgiles — CHANGES_REQUESTED** [review]
Oh man, this is so much better!  Seems to work well locally and in Azure.  I left a few comments below.

File: infrastructure/conf/nginx-conf-local/default [hunk 1]
```diff
@@ -20,47 +20,37 @@ server {
...
+    }
+    location = /fr/communities/digital-talent {
+        absolute_redirect on;
+        return 301 https://www.canada.ca/fr/gouvernement/systeme/gouvernement-numerique/collectivite-gcnumerique.html;
+    }
+    location = /en/talent/search { return 301 /en/search$is_args$args; }
+    location = /fr/talent/search { return 301 /fr/search$is_args$args; }
+    location = /en/applications { return 301 /en/applicant/profile-and-applications$is_args$args; }
+    location = /fr/applications/ { return 301 /fr/applicant/profile-and-applications$is_args$args; }
+    location = /en/applications/ { return 301 /en/applicant/profile-and-applications$is_args$args; }
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply infrastructure/conf/nginx-conf-local/default:45]: Yep

File: infrastructure/conf/nginx-conf-deploy/default [hunk 2]
```diff
@@ -10,47 +10,32 @@ server {
...
 
     # Permanent redirect for IAP french vanity url
-    location = /apprentis-autochtone-ti {
-        absolute_redirect off;
-        return 301 /fr/indigenous-it-apprentice;
-    }
-
-    # enable serving gzip-ed files
-    gzip_static on;
+    location = /apprentis-autochtone-ti { return 301 /fr/indigenous-it-apprentice; }
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply infrastructure/conf/nginx-conf-deploy/default:38]: Good point, no reason not to have it here as well

File: infrastructure/conf/nginx-conf-local/default [hunk 3]
```diff
@@ -74,45 +64,71 @@ server {
         break;
     }
 
-    # enable serving gzip-ed files
-    gzip_static on;
+    # Disable .git directory
+    location ~ /\.git {
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply infrastructure/conf/nginx-conf-local/default:68]: For sure, that was an oversight!

File: infrastructure/conf/nginx-conf-deploy/default [hunk 4]
```diff
@@ -10,47 +10,32 @@ server {
...
-    }
+    location = /en/communities/digital-talent {
+        absolute_redirect on;
+        return 301 https://www.canada.ca/en/government/system/digital-government/gcdigital-community.html;
+    }
+    location = /fr/communities/digital-talent {
+        absolute_redirect on;
+        return 301 https://www.canada.ca/fr/gouvernement/systeme/gouvernement-numerique/collectivite-gcnumerique.html;
+    }
+    location = /en/talent/search { return 301 /en/search$is_args$args; }
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply infrastructure/conf/nginx-conf-deploy/default:28]: Do you think I should do that in this PR or make another ticket?

**Review by petertgiles — APPROVED** [review]
Looks great!  Thanks for the fixups.

File: infrastructure/conf/nginx-conf-deploy/default (hunk 4 above)
[the remark this answers is not in the document]
Comment (petertgiles) [reply infrastructure/conf/nginx-conf-deploy/default:28]: It's probably small enough to include it but today being a cutoff - let's merge this now and spin it off.

**Pull request merged**